test/k23si/3SITxnTest.cpp

```diff
@@ -1279,6 +1285,484 @@ seastar::future<> testScenario01() {
+                    .then([&](auto&& response) {
+                        auto& [status, resp] = response;
+                        K2INFO("SC02.case07()::OP_READ. " << "status: " << status.code << " with MESG: " << status.message);
+                        K2INFO("Value of Key5: " << resp.value.val);
+                        K2EXPECT(status, Statuses::S200_OK);
+                        K2EXPECT(resp.value.val, v3);
+                    });
+                });
+            }
+        );
+    }) // end sc-02 case-07
+    .then([] {
+        // SC02 case08 & 09 & 10: READ existing key at time before/equal/after the key
+        return getTimeNow();
+    })
+    .then([this](dto::Timestamp&& ts) {
+        std::cout << std::endl;
+        K2INFO("------- SC02.case08 & 09 & 10 (READ existing key at time before/equal/after the key) -------");
+
+        return seastar::do_with(
+            dto::K23SI_MTR {.txnid = txnids++, .timestamp = std::move(ts), .priority = dto::TxnPriority::Medium},
+            dto::Key {.partitionKey = "SC02_pkey8", .rangeKey = "range8"},
+            DataRec {.f1="SC02_f08", .f2="SC02_f09"},
+            [this](auto& eqMTR, auto& k6, auto& v1) { 
+                return doWrite<DataRec>(k6, v1, eqMTR, k6, collname, false, true, ErrorCaseOpt::NoInjection)
+                .then([](auto&& response) {
+                     auto& [status, resp] = response;
+                     K2EXPECT(status, Statuses::S201_Created);               
+                     K2INFO("SC02.case08-09-10(write an unread key)::OP_Write. " << "status: " << status.code << " with MESG: " << status.message);
+                })
+                .then([this, &k6, &eqMTR] {
+                    return doEnd(k6, eqMTR, collname, true, {k6}, Duration(0us), ErrorCaseOpt::NoInjection)
+                    .then([](auto&& response) {
+                        auto& [status, resp] = response;
+                        K2EXPECT(status, Statuses::S200_OK);               
+                        K2INFO("SC02.case08-09-10(end an unread key)::OP_END. " << "status: " << status.code << " with MESG: " << status.message);
+                    });
+                })
+                .then([this, &k6, &eqMTR] {
+                    dto::K23SI_MTR bfMTR{
+                        .txnid = txnids++, 
+                        .timestamp = {(eqMTR.timestamp.tEndTSECount() - 500000000), 123, 1000}, // 500ms earlier
+                        .priority = dto::TxnPriority::Medium};
+                    return doRead<DataRec>(k6, bfMTR, collname, ErrorCaseOpt::NoInjection)
+                    .then([](auto&& response) {
+                        auto& [status, resp] = response;
+                        K2INFO("SC02.case08-09-10(before)::OP_READ. " << "status: " << status.code << " with MESG: " << status.message);
+                        K2EXPECT(status, Statuses::S404_Not_Found);               
```

Comment:
Here and elsewhere you can use the statuses from K23SIStatus in dto/K23SI.h which is a bit more clear in intent